merge master

Nic-Ma opened the pull request:
Fixes # .

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
Hi @wyli ,

Did you miss some modification in this PR? I can't find the  "Feature highlights" change.
Thanks.

wyli commented:
> Hi @wyli ,
> 
> Did you miss some modification in this PR? I can't find the "Feature highlights" change.
> Thanks.

thanks I updated it

wyli commented:
@Nic-Ma in the module figure, we could make the research block more generic: 
`Research - implementation of recently published deep learning research in medical imaging, such as COVID-19 CT models, automatic model parallelism.`

Nic-Ma commented:
> @Nic-Ma in the module figure, we could make the research block more generic:
> `Research - implementation of recently published deep learning research in medical imaging, such as COVID-19 CT models, automatic model parallelism.`

Sure, updated in the latest commit.
Thanks.

wyli commented:
> > @Nic-Ma in the module figure, we could make the research block more generic:
> > `Research - implementation of recently published deep learning research in medical imaging, such as COVID-19 CT models, automatic model parallelism.`
> 
> Sure, updated in the latest commit.
> Thanks.

could you please update the boxes in the figure? "COVID-19" should be more specific, something like "COVID-19 CT segmentation". And not sure if we want to put FL in the box, please align with Prerna.

Nic-Ma commented:
Sure, updated in the latest commit.
Thanks.

wyli commented:
> Sure, updated in the latest commit.
> Thanks.

just saw that you add new figures, but I didn't generate those images, they were downloaded from the internet, I'm not sure if we can use them freely. please generate new ones, we just need some label images. https://github.com/Nic-Ma/MONAI/blob/add-arch-chart-docs/docs/source/highlights.md#8-post-processing-transforms-for-model-output

Nic-Ma commented:
OK, I am working on the `DecathlonDataset`, will try to update the post transform chart with Spleen data soon.

Thanks.

Nic-Ma commented:
Hi @wyli , updated the chart in the latest commit.
Thanks.

wyli committed:
Enhance csrc/resample (#1471)

* update openmp flag

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* improves boundtype docs

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* update setup.py

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* input validation 1d

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* fixes typos

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* fixes typos

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* merge upstream changes

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* tests enums

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* init. test grid pull

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* update

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* test grid_pull

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* fixes min test

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* adds device tests

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* bwd tests

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

rijobro committed:
move transforms out of compose file (#1623)

move transforms out of compose file. this is backwards compatible.

rijobro committed:
pad_list_data_collate (#1626)

* pad_collation

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

* increase number of test cases to ensure required testing errors

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

* determinism in setUp

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

* pad collate for list of lists

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

* code format

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

* allow padding options

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

rijobro committed:
decollate batch (#1624)

decollate batch

rijobro committed:
enhance random range parameters (#1635)

enhance random range parameters

If element `i` is iterable, then `uniform[-rotate_range[i][0], rotate_range[i][1])` will be used to generate the rotation parameter for the ith dimension. If not, `uniform[-rotate_range[i], rotate_range[i])` will be used. This can be altered on a per-dimension basis. E.g., `((0,3), 1, ...)`: for dim0, rotation will be in range `[0, 3]`, and for dim1 `[-1, 1]` will be used. Setting a single value will use `[-x, x]` for dim0 and nothing for the remaining dimensions.

Backwards compatibility is ensured because the old behaviour used to do `uniform[-rotate_range[i], rotate_range[i])`, and this PR just adds the possibility to do `uniform[-rotate_range[i][0], rotate_range[i][1])`.

ericspod committed:
Loss script fixes (#1645)

* Fixing issues preventing loss functions from being compatible with Torchscript

Signed-off-by: Eric Kerfoot <eric.kerfoot@kcl.ac.uk>

* Updates

Signed-off-by: Eric Kerfoot <eric.kerfoot@kcl.ac.uk>

* Updates

Signed-off-by: Eric Kerfoot <eric.kerfoot@kcl.ac.uk>

* [MONAI] python code formatting

Signed-off-by: monai-bot <monai.miccai2019@gmail.com>

* Updates

Signed-off-by: Eric Kerfoot <eric.kerfoot@kcl.ac.uk>

* Adding conditional skip to Torchscript tests

Signed-off-by: Eric Kerfoot <eric.kerfoot@kcl.ac.uk>

Co-authored-by: monai-bot <monai.miccai2019@gmail.com>

Nic-Ma committed:
1633 fix string length > 1024 issue in all_gather (#1634)

* [DLMED] fix length > 1024 issue in string list all gather

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] fix flake8 issue

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] update according to comments

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] update according to comments

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] add more test

Signed-off-by: Nic Ma <nma@nvidia.com>

Co-authored-by: Yiheng Wang <68361391+yiheng-wang-nv@users.noreply.github.com>

leotam committed:
1567 add RemoveRepeatedChannel (#1569)

* DeleteChannel array, dictionary, tests

Signed-off-by: Leo Tam <leot@nvidia.com>

* formatting

Signed-off-by: Leo Tam <leot@nvidia.com>

* DeleteChannel array, dictionary, tests

Signed-off-by: Leo Tam <leot@nvidia.com>

* Register in transform init

Signed-off-by: Leo Tam <leot@nvidia.com>

* clean comment

Signed-off-by: Leo Tam <leot@nvidia.com>

* Dict init

Signed-off-by: Leo Tam <leot@nvidia.com>

* Register D and Dict

Signed-off-by: Leo Tam <leot@nvidia.com>

* Style fix

Signed-off-by: Leo Tam <leot@nvidia.com>

* Safety cast

Signed-off-by: Leo Tam <leot@nvidia.com>

* Changinging names from DeleteChannel to RemoveRepeatedChannel

Signed-off-by: Leo Tam <leot@nvidia.com>

* Lint fixes

Signed-off-by: Leo Tam <leot@nvidia.com>

* Test import fix, autofix again

Signed-off-by: Leo Tam <leot@nvidia.com>

Nic-Ma committed:
[DLMED] add progress arg (#1647)

Signed-off-by: Nic Ma <nma@nvidia.com>

Co-authored-by: Eric Kerfoot <17726042+ericspod@users.noreply.github.com>

wyli committed:
1650-update distcall (#1652)

* update distcall

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* update final closing

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Co-authored-by: Eric Kerfoot <17726042+ericspod@users.noreply.github.com>

bhashemian committed:
Fix ToPIL (#1648)

* Add ToPIL transformation

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add ToPILd, ToPILD, ToPILDict

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Remove has_pil

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Include ToPIL, ToPILd, ToPILD, and ToPILDict

Also include ToNumpyD, ToNumpyDict, TorchVisionD, and TorchVisionDict

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix a typing issue

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix PIL optional import

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add unittests for ToPIL and ToPILD

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix formatting

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix formatting

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add PILImage.Image as the input for ToNumpy and ToTensor

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix type checking issue for PIL.Image.Image

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Change PILImage_fromarray to lower case

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

rijobro committed:
Randomizable constructor (#1639)

add RandomizableTransform

YuanTingHsieh committed:
Add more tests to test_deepgrow_dataset (#1654)

* Add more tests to test_deepgrow_dataset

Signed-off-by: YuanTingHsieh <yuantinghsieh@gmail.com>

YuanTingHsieh committed:
Add Deepgrow inference transforms (#1653)

* Add inference transforms

Signed-off-by: YuanTingHsieh <yuantinghsieh@gmail.com>

* Remove unused import

Signed-off-by: YuanTingHsieh <yuantinghsieh@gmail.com>

* Fix review comments

Signed-off-by: YuanTingHsieh <yuantinghsieh@gmail.com>

Co-authored-by: SACHIDANAND ALLE <sachidanand.alle@gmail.com>

Nic-Ma committed:
1668 Add RandAxisFlip transforms (#1670)

* [DLMED] add RandAxisFlip transforms

Signed-off-by: Nic Ma <nma@nvidia.com>

* [MONAI] python code formatting

Signed-off-by: monai-bot <monai.miccai2019@gmail.com>

* [DLMED] fix flake8 issues

Signed-off-by: Nic Ma <nma@nvidia.com>

Co-authored-by: monai-bot <monai.miccai2019@gmail.com>

rijobro committed:
allow dictionary image only and test endianness (#1669)

fix big endianness problem for dictionary reading. Also allow dictionary image_only

Nic-Ma committed:
1673 Add AutoAdjustChannel transform (#1675)

* [DLMED] add AutoAdjustChannel transform

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] add dict version transform

Signed-off-by: Nic Ma <nma@nvidia.com>

* [MONAI] python code formatting

Signed-off-by: monai-bot <monai.miccai2019@gmail.com>

* [DLMED] fix doc-build issue

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] fix flake8 issue

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] fix flake8 issue

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] update according to Wenqi's comments

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] update doc-strings

Signed-off-by: Nic Ma <nma@nvidia.com>

* [MONAI] python code formatting

Signed-off-by: monai-bot <monai.miccai2019@gmail.com>

Co-authored-by: monai-bot <monai.miccai2019@gmail.com>

rijobro committed:
[1662] Allow missing keys (#1676)

Allow missing keys

rijobro committed:
nifti saver squeeze dims (#1680)

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

rijobro committed:
improve gradcam doc (#1682)

Signed-off-by: Richard Brown <33289025+rijobro@users.noreply.github.com>

wyli committed:
fixes integration tests (#1683)

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

yiheng-wang-nv committed:
Update load pretrain for densenet (#1684)

* Update load pretrain for densenet

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

* Fix isort issue

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

wyli committed:
revise the dtype according to the discussion (#1671)

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

yiheng-wang-nv committed:
Fix senet pretrained weights issue (#1689)

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

Nic-Ma committed:
1680 Add squeeze SegmentationSaver handler (#1686)

* [DLMED] add suqeeze to handler

Signed-off-by: Nic Ma <nma@nvidia.com>

* [DLMED] update according to comments

Signed-off-by: Nic Ma <nma@nvidia.com>

wyli committed:
1285 adds module list (#1690)

* fixes #1285

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* adds test

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

wyli committed:
1685-upgrade base image to 21.02 (#1687)

* fixes #1685

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* add temp test

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* adds docstring

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* fixes dist sampler

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* remove temp tests

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* fixes type hint issue

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

wyli committed:
fixes data type in switching (#1694)

* fixes data type in switching

Signed-off-by: Wenqi Li <wenqil@nvidia.com>

* [MONAI] python code formatting

Signed-off-by: monai-bot <monai.miccai2019@gmail.com>

Co-authored-by: monai-bot <monai.miccai2019@gmail.com>

bhashemian committed:
WSI reader (#1548)

* Implement CuImageReader and OpenSlideReader

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add unittests for CuImageReader

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add unittests for OpenSlideReader

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Sort imports

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add correct boundaries

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add test cases for reading patches on a grid for CuImage

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add patch whole slide imaging dataset for pathology

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add test case for read patches for OpenSlide

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* flake8 and few minor changes

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* black

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* flake8

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add kwargs to CuImageReader and OpenSlideReader's read method

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Change the type hint from np.dtype to DTypeLike

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix a bug

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Implement WSIReader and unittests

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Minor updates

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix few typing issues

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Revert datasets

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add shape property to openslide image object
Reverse size to be compatible with output size (hxw)

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add untittest for loading the whole image
Reverse the size accroding to the WSIReader

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Update the whole image size

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Remove optional size

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Remove optional dtype

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Remove _get_spatial_shape return type

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Reverse the orders of dimensions of `location`
to be compatible with image shape

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Change test cases to use smaller image and revese location's dimensions

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Replace the test TIFF and some upgrades

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Update dependencies for OpenSlide

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Update unittests for OpenSlide and CuImage

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix openslide dependency

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Fix doc dependencies

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Minor changes

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Few variable name changes

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add EnsureChannelFirst

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

* Add metadata to WSIReader

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>

Nic-Ma committed:
[DLMED] add WSIReader to LoadImage (#1695)

Signed-off-by: Nic Ma <nma@nvidia.com>

bhashemian committed:
Raise import error for openslide and cuimage (#1699)

Signed-off-by: Behrooz <3968947+behxyz@users.noreply.github.com>
